### ISSUE TITLE
Error in sample.int(length(x), size, replace, prob) when using leidenalg on a relatively big matrix 

### ISSUE THREAD

**Issue by kaizen89** [satijalab/seurat#2431]
I am trying to use leidenalg in FindClusters function. The function is running correctly on small dataset like `pbmc_small`
However, when running
```
dim(object)
[1] 23097 37996
object <-  FindClusters(object, resolution = 0.2, algorithm = 4, method="igraph") 
```
it throws an error
```
Error in sample.int(length(x), size, replace, prob) : 
  invalid first argument
In addition: Warning message:
In max(connectivity, na.rm = T) :
  no non-missing arguments to max; returning -Inf
```
```
traceback()
7: sample.int(length(x), size, replace, prob)
6: sample(x = names(x = connectivity[mi]), 1)
5: GroupSingletons(ids = ids, SNN = object, group.singletons = group.singletons, 
       verbose = verbose)
4: FindClusters.default(object = object[[graph.name]], modularity.fxn = modularity.fxn, 
       initial.membership = initial.membership, weights = weights, 
       node.sizes = node.sizes, resolution = resolution, method = method, 
       algorithm = algorithm, n.start = n.start, n.iter = n.iter, 
       random.seed = random.seed, group.singletons = group.singletons, 
       temp.file.location = temp.file.location, edge.file.name = edge.file.name, 
       verbose = verbose, ...)
3: FindClusters(object = object[[graph.name]], modularity.fxn = modularity.fxn, 
       initial.membership = initial.membership, weights = weights, 
       node.sizes = node.sizes, resolution = resolution, method = method, 
       algorithm = algorithm, n.start = n.start, n.iter = n.iter, 
       random.seed = random.seed, group.singletons = group.singletons, 
       temp.file.location = temp.file.location, edge.file.name = edge.file.name, 
       verbose = verbose, ...)
2: FindClusters.Seurat(object, resolution = 0.2, algorithm = 4)
1 : FindClusters(object, resolution = 0.2, algorithm = 4)
```
I was able to make it run on my object by subsampling both cells and genes.
Is there any solution to this issue which does not seem to be related to `leidenalg` ?

**Comment by chlee-tabin** [comment]
I think it is the same issue I had:
https://github.com/satijalab/seurat/issues/2438#issue-539325312

Either reverting back to 3.1.1 or applying the pull request on the linked issue solved the problem for me.

**Comment by kaizen89** [comment]
Reverting back to version 3.1.1 did not solve the issue, rstudio crashes probably as I have a big matrix, when setting `method="igraph" ` I have the error mentionned above.

**Comment by TomKellyGenetics** [comment]
@kaizen89 do you get the same error when using `method = "matrix`? This is the default (recommended) setting unless you have a very large dataset.

For troubleshooting purposes please try running it with  `group.singletons = FALSE` and run `table(object@active.ident)` to show the raw clusters.

**Comment by Pestudkaru** [comment]
I have the same problem with a small data set. If I set group.singletons = FALSE then it runs. I get 1098 singletons from table().
How can I fix the issue?

**Comment by TomKellyGenetics** [comment]
@Pestudkaru Which version of Seurat are you using? PR #2440 should be merged into 3.1.3 now.

**Comment by kaizen89** [comment]
Leidenalg is working now using the latest version. The only problem is that I get 3 clusters with reso=1 instead of at least 9 clusters with louvain with reso=0.4. I was not expecting that.

**Comment by TomKellyGenetics** [comment]
@kaizen89 In my understanding, the resolution parameter is only supported in FindClusters for the Leiden algorithm, not for Louvain. Please bear in mind that results of the algorithm will be different as described by Traag _et al_ (2019). Which version of the Louvain algorithm are you using? `algorithm = 3` should give similar results as this comes from the same research group as Leiden and has several of the same refinements.

**Comment by kaizen89** [comment]
The resolution is definitely supported by Louvain ( changing the resolution will change the number of clusters).
I get similar results between Louvain and SLM algorithm, but really different with Leiden (low number of clusters even at higher resolutions).

**Comment by Pestudkaru** [comment]
> @Pestudkaru Which version of Seurat are you using? PR #2440 should be merged into 3.1.3 now.

It's working now! I also updated to the newest 3.1.4. version from Cran.

**Comment by TomKellyGenetics** [comment]
@kaizen89 The Leiden algorithm passes arguments to the python version "leidenalg". It should give the same results as that. Another possible explanation is that weighted graph objects are supported by Leiden in recently releases (this may change the default results of FindClusters are discussed here #2574). Not sure if Louvain supports weights in Suerat which may explain this difference.
